src/main/scala/is/hail/variant/Genotype.scala

```diff
@@ -566,24 +468,245 @@ object Genotype {
+      require(_gt == Genotype.gtFromLinear(_px).getOrElse(-1))
+    }
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case g: GenericGenotype =>
+      _gt == g._gt &&
+        ((_ad == null && g._ad == null)
+          || (_ad != null && g._ad != null && _ad.sameElements(g._ad))) &&
+        _dp == g._dp &&
+        _gq == g._gq &&
+        fakeRef == g.fakeRef &&
+        isDosage == g.isDosage &&
+        ((_px == null && g._px == null)
+          || (_px != null && g._px != null && _px.sameElements(g._px)))
+
+    case _ => false
+  }
+
+  override def hashCode: Int =
+    new HashCodeBuilder(43, 19)
+      .append(_gt)
+      .append(util.Arrays.hashCode(_ad))
+      .append(_dp)
+      .append(_gq)
+      .append(util.Arrays.hashCode(_px))
+      .append(fakeRef)
+      .append(isDosage)
+      .toHashCode
+
+  def unboxedGT: Int = _gt
+
+  def gt: Option[Int] =
+    if (_gt >= 0)
+      Some(_gt)
+    else
+      None
+
+  def ad: Option[Array[Int]] = Option(_ad)
+
+  def dp: Option[Int] =
+    if (_dp >= 0)
+      Some(_dp)
+    else
+      None
+
+  def od: Option[Int] =
+    if (_dp >= 0 && _ad != null)
+      Some(_dp - _ad.sum)
+    else
+      None
+
+  def gq: Option[Int] =
+    if (_gq >= 0)
+      Some(_gq)
+    else
+      None
+
+  def px: Option[Array[Int]] = Option(_px)
+
+  def pl: Option[Array[Int]] =
+    if (_px == null)
+      None
+    else if (!isDosage)
+      Some(_px)
+    else
+      Some(Genotype.linearToPhred(_px))
+
+  def dosage: Option[Array[Double]] =
+    if (_px == null)
+      None
+    else if (isDosage)
+      Some(_px.map(_ / 32768.0))
+    else
+      Some(Genotype.phredToDosage(_px))
+
+  override def toString: String = {
+    val b = new StringBuilder
+
+    b.append(gt.map { gt =>
+      val p = Genotype.gtPair(gt)
+      s"${ p.j }/${ p.k }"
+    }.getOrElse("./."))
+
+    if (fakeRef) {
+      b += '*'
+    }
+
+    b += ':'
+    b.append(ad.map(_.mkString(",")).getOrElse("."))
+    b += ':'
+    b.append(dp.map(_.toString).getOrElse("."))
+    b += ':'
+    b.append(gq.map(_.toString).getOrElse("."))
+    b += ':'
+    if (!isDosage) {
+      b.append("PL=" + pl.map(_.mkString(",")).getOrElse("."))
+    } else {
+      b.append("GP=" + dosage.map(_.mkString(",")).getOrElse("."))
+    }
+
+    b.result()
+  }
+
+  def toRow: Row =
+    Row(gt.orNull, ad.map(_.toSeq).orNull, dp.orNull, gq.orNull, px.map(_.toSeq).orNull, fakeRef, isDosage)
+
+  def toJSON: JValue = JObject(
+    ("gt", gt.map(JInt(_)).getOrElse(JNull)),
+    ("ad", ad.map(ads => JArray(ads.map(JInt(_)).toList)).getOrElse(JNull)),
+    ("dp", dp.map(JInt(_)).getOrElse(JNull)),
+    ("gq", gq.map(JInt(_)).getOrElse(JNull)),
+    ("px", px.map(pxs => JArray(pxs.map(JInt(_)).toList)).getOrElse(JNull)),
+    ("fakeRef", JBool(fakeRef)),
+    ("isDosage", JBool(isDosage)))
+}
+
+object GenericGenotype {
+  def read(nAlleles: Int, isDosage: Boolean, a: ByteIterator): GenericGenotype = {
```

Comment:
I probably would have left read on Genotype.  Next thing you're going to do is make it return a MutableGenotype.  The client of this code doesn't care how the returned genotypes are implemented.